orchagent/fgnhgorch.cpp

```diff
@@ -0,0 +1,1379 @@
+        }
+        else 
+        {
+            /* delete all associated FG_NHG and SAI objects */
+            if (fgNhg_entry->second.prefixes.size() == 0 && fgNhg_entry->second.nextHops.size() == 0)
+            {
+                m_FgNhgs.erase(fgNhg_entry);
+                assert(m_FgNhgs.find(fgNhg_name) == fgNhgPrefixes.end());
+                SWSS_LOG_INFO("%s: Received delete call for valid entry with no further dependencies, deleting %s",
+                        __FUNCTION__, fgNhg_name.c_str());
+            }
+            else
+            {
+                SWSS_LOG_NOTICE("%s: still prefixes and nexthop in FG_NHG group %s", __FUNCTION__, fgNhg_name.c_str());
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+
+bool FgNhgOrch::doTaskFgNhg_prefix(const KeyOpFieldsValuesTuple & t)
+{
+    SWSS_LOG_ENTER();
+    string op = kfvOp(t);
+    string key = kfvKey(t);
+    IpPrefix ip_prefix = IpPrefix(key);
+    auto prefix_entry = fgNhgPrefixes.find(ip_prefix);
+    bool route_handled = false;
+
+    if (op == SET_COMMAND)
+    {
+        if(prefix_entry != fgNhgPrefixes.end())
+        {
+            SWSS_LOG_INFO("%s FG_NHG prefix already exists", __FUNCTION__);
+            return true;
+        }
+
+        string fgNhg_name = "";
+        for (auto i : kfvFieldsValues(t))
+        {
+            if (fvField(i) == "FG_NHG")
+            {
+                fgNhg_name = fvValue(i);
+            }
+        }
+        if(fgNhg_name == "")
+        {
+            SWSS_LOG_ERROR("Received FG_NHG with empty name for key %s", kfvKey(t).c_str());
+            return true;
+        }
+
+        auto fgNhg_entry = m_FgNhgs.find(fgNhg_name);
+        if(fgNhg_entry == m_FgNhgs.end())
+        {
+            SWSS_LOG_INFO("%s FG_NHG entry not received yet, continue", __FUNCTION__);
+            return false;
+        }
+        else 
+        {
+            /* delete regular ecmp handling for prefix */
+            sai_object_id_t vrf_id = gVirtualRouterId;
+            auto route_table_entry = gRouteOrch->getSyncdRoutes().find(vrf_id);
+            NextHopGroupKey nhg;
+            if (route_table_entry == gRouteOrch->getSyncdRoutes().end())
+            {
+                SWSS_LOG_INFO("Failed to find route table, vrf_id 0x%lx\n", vrf_id);
+            }
+            else
+            {
+                /* ipprefix already exist in routeorch */
+                auto it_route = route_table_entry->second.find(ip_prefix);
+                if (it_route != route_table_entry->second.end())
+                {
+                    nhg = it_route->second;
+                    RouteBulkContext ctx;
+                    ctx.ip_prefix = ip_prefix;
+                    ctx.vrf_id = vrf_id;
+                    ctx.nhg = nhg;
+                    gRouteOrch->removeRoute(ctx);
+                    gRouteOrch->gRouteBulker.flush();
+                    gRouteOrch->removeRoutePost(ctx);
```

Comment:
How about the failures here? If this fails, it will not be removed from route_table and you'll end up adding fgnhg prefix as well.